### DISCUSSION TITLE
[V5] Consuming IoC services from within services

### DISCUSSION THREAD

**Discussion by milgner** [adonisjs/core#1908] [Help]
So I started building a small application with Adonis.js and so far everything has been really nice!
There's one thing I can't seem to wrap my head around, though:

I have a service that provides a function to create a URL. I'm injecting it into a controller and a listener and everything works fine. Now however I started enhancing another service - which by itself can also be injected just fine - in a way that would also require it to use this URL-generation service.

As soon as I add the ioc-import (at the end of my imports), the application stops build and complains about
```
IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Addons/Mail" namespace from the IoC Container
```
The Mail import is my second import, though, even before the newly-added import. Are there any recommendations / best practices for this kind of scenario? I think that using services from other services is not that uncommon?

**Comment by rrmesquita** [comment]
I got a similar problem with the IoC. I'm declaring a contract and trying to import it into a command file.

```ts
// contracts/cron.ts

declare module '@ioc:Adonis/Addons/Cron' {
  abstract class CronJob {
    // ...
  }
}
```

```ts
import { CronJob } from '@ioc:Adonis/Addons/Cron'
```

But I'm having the same output message as @milgner :

```sh
IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Addons/Cron" namespace from the IoC Container
```

**Comment by Banashek** [comment]
Could you both post more of your code?

Remember that the items are not available from raw imports (since those imports happen when the file is read by node), but rather you have to use the appropriate `ios.use` method to get the actual service instance in your files.

To be more specific, make sure you're using the `IoC` methods similar to the [v4 docs](https://adonisjs.com/docs/4.1/ioc-container#_resolving_dependencies) rather than trying to raw `import` your service and use it that way. Things instantiated by the `IoC` container need to be retrieved from there as well.

**Reply by milgner** [reply]
I think these might be two separate issues. Since I already successfully created one service and was under the impression that I understood how things worked, I now did a bit of divide-and-conquer on my setup it turns out that the error is different than I first thought. Effectively it boiled down to this:

Everything works fine up to the point where, inside my service, I import *and use* 
```
import Mail from '@ioc:Adonis/Addons/Mail'
```
(I guess if the import is not used in the file that it is optimized away before it can cause issues.)
I tried moving around the order of imports but that doesn't seem to have any effect.

Here's the full output:

```
IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Addons/Mail" namespace from the IoC Container

at /project/app/Services/LoginMailService.ts(anonymous):6

1 Function.lookupFailed
  /project/node_modules/@adonisjs/fold/build/src/Exceptions/IocLookupException.js:18

2 Ioc.onLookupFailed
  /project/node_modules/@adonisjs/fold/build/src/Ioc/index.js:61

3 Ioc.use
  /project/node_modules/@adonisjs/fold/build/src/Ioc/index.js:401

4 Module._compile
  /project/node_modules/pirates/lib/index.js:99

5 Module._compile
  /project/node_modules/pirates/lib/index.js:99

6 newLoader
  /project/node_modules/pirates/lib/index.js:104

7 Object.newLoader [as .ts]
  /project/node_modules/pirates/lib/index.js:104


IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Addons/Mail" namespace from the IoC Container

at /project/app/Services/LoginMailService.ts(anonymous):3
1
2  //import Config from '@ioc:Adonis/Core/Config'
3  import Mail from '@ioc:Adonis/Addons/Mail'
4  //import UrlAuthService from '@ioc:MyApp/UrlAuthService'
5  import { LoginMailServiceInterface } from '@ioc:MyApp/LoginMailService'
6
7  import User from 'App/Models/User'
8

1 Function.lookupFailed
  /project/node_modules/@adonisjs/fold/build/src/Exceptions/IocLookupException.js:18

2 Ioc.onLookupFailed
  /project/node_modules/@adonisjs/fold/build/src/Ioc/index.js:61

3 Ioc.use
  /project/node_modules/@adonisjs/fold/build/src/Ioc/index.js:401

4 Module._compile
  /project/node_modules/pirates/lib/index.js:99

5 Object.newLoader [as .ts]
  /project/node_modules/pirates/lib/index.js:104

IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Lucid/Orm" namespace from the IoC Container
    at Function.lookupFailed (/project/node_modules/@adonisjs/fold/build/src/Exceptions/IocLookupException.js:18:16)
    at Ioc.onLookupFailed (/project/node_modules/@adonisjs/fold/build/src/Ioc/index.js:61:59)
    at Ioc.use (/project/node_modules/@adonisjs/fold/build/src/Ioc/index.js:401:18)
    at Object.<anonymous> (/project/app/Models/User.ts:2:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Module._compile (/project/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.newLoader [as .ts] (/project/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
```

Note that I can send mails with the `@ioc:Adonis/Addons/Mail` import just fine if I move it out of the service and back into my Listener. It only causes problems when importing it from within my service. But I thought that it would be nicer to have a service for this instead of cluttering my controller or listener with details about mailing logic so I want to move this stuff into a service.

**Reply by milgner** [reply]
Just had another go at this because I notized that the file containing my service definition was, alphabetically sorted, before the `mail.ts` with the mail service that's being imported here. Unfortunately, renaming it to `zzz_myservice.ts` did not have any effect either.

**Comment by RomainLanz** [comment]
Hey all of you! 👋 

Could you please post some code of what you are doing?

Also, take notes that if you are doing a Service Provider, this behavior is intended, you shouldn't rely on framework internals when building a service provider.

**Comment by prescience-data** [comment]
(I think) I finally figured this out fwiw:

## 1. No non-Core IOC access inside the `Service`!

~~Make sure your `Service` does not import anything from the `@ioc:<YOUR_NAMESPACE>` namespace, as it is not ready yet.~~
Update: I didn't understand this properly initially due to still learning how TS modules work, it should be find provided your `Contracts` defs only create a reference to the final injected dependency rather than importing directly as one guide I found on google instructed.
Following the actual Adonis packages this is how it seems to work...

```typescript

declare module "@ioc:MyApp/Foo" {
  export interface FooServiceContract {
 
    getItem(itemName: FooName): FooDriver
  }

  const Foo: FooServiceContract
  export default Foo
}
```



## 2. Avoid custom TypeScript aliases in your `Providers`!

If you are defining custom aliases such as `Service/Foo` - when importing within your `Provider`, make sure to import the `Service` directly rather than TypeScript alias as this fails on build for some reason: 

```typescript
import Auditor from "Services/Auditor
// should be 
import Auditor from "../services/Auditor"
```

## 3. Load `Providers` in order of priority!

Make sure you arrange the boot order of your Service Providers in your `.adonisrc.json` file in order of priority, ie if `@ioc:MyApp/Dog` requires `@ioc:MyApp/Ball` somewhere in it's logic, make sure to load `Ball` ahead of `Dog`:

```json
{
  "providers": [
    "./providers/AppProvider",
    "@adonisjs/core",
    "./providers/BallProvider",
    "./providers/DogProvider"
  ],
}
```

## 4. Finally, remember `public static needsApplication = true`!

This one is crucially frustrating if you are having these build problems: Make sure to add the `public static needsApplication = true` to your `Provider` if it requires the app!

```typescript
import { ApplicationContract } from "@ioc:Adonis/Core/Application"
import FooService from "../services/FooService"

/**
 * @class FooProvider
 * @classdesc Provides the Foo service to the application.
 */
export default class FooProvider {
  public static needsApplication = true // THIS CAUSED ME SO MUCH PAIN!!! 😫😫😫

  constructor(protected app: ApplicationContract) {}

  public register(): void {
    this.app.container.singleton("MyApp/Foo", () => {
      return new FooService()
    })
  }
}
```

**Reply by jcjp** [reply]
Didn't worked for me what am I missing?

```javascript
// contracts/interfaces/User.interface.ts
declare module '@ioc:Project/UserInterface' {
  export default interface UserInterface {
    ...
  }
}
```

```javascript
// providers/User/index.ts
import UserInterface from '@ioc:Project/UserInterface'
export default class UserService implements UserInterface {
  ...
}
```

```javascript
// providers/AppProvider.ts
import {ApplicationContract} from '@ioc:Adonis/Core/Application'
import UserService from './User';
export default class AppProvider {
  constructor (protected app: ApplicationContract) {
  }

  public register () {
    this.app.container.bind('./UserService', () => {
      const Config = this.app.container.use('Adonis/Src/Config');
      return new UserService(Config)
    });
  }
  // no values for other functions
}
```

Also tried your solution above still encountered the error.

**Reply by 4konung** [reply]
You should import your UserService like [this](https://docs.adonisjs.com/guides/application#-move-import-inside-the-boot-method)
```javascript
import { ApplicationContract } from "@ioc:Adonis/Core/Application"

/**
 * @class FooProvider
 * @classdesc Provides the Foo service to the application.
 */
export default class FooProvider {
  public static needsApplication = true // THIS CAUSED ME SO MUCH PAIN!!! 😫😫😫

  constructor(protected app: ApplicationContract) {}

  public register(): void {
  const { default: FooService } =  await import("../services/FooService")
  
    this.app.container.singleton("MyApp/Foo", () => {
      return new FooService()
    })
  }
}
```

But this will not fix the problem if you import in your FooService another service that was made like this one (FooService).

**Comment by danielzesteves** [comment]
Hi

This works for me, resolving the service in the provider.


![imagen](https://user-images.githubusercontent.com/12122895/194784606-87eb8115-0858-422d-91eb-9799a73bfe2d.png)



![imagen](https://user-images.githubusercontent.com/12122895/194784649-a62324ea-b1b9-4525-92bc-684e32dd9fd6.png)
